src/main/java/jcifs/dcerpc/msrpc/lsarpc.java

```diff
@@ -0,0 +1,1046 @@
+                this.security_quality_of_service.decode(_src);
+
+            }
+        }
+    }
+
+    public static class LsarDomainInfo extends NdrObject {
+
+        public rpc.unicode_string name;
+        public rpc.sid_t sid;
+
+
+        @Override
+        public void encode ( NdrBuffer _dst ) throws NdrException {
+            _dst.align(4);
+            _dst.enc_ndr_short(this.name.length);
+            _dst.enc_ndr_short(this.name.maximum_length);
+            _dst.enc_ndr_referent(this.name.buffer, 1);
+            _dst.enc_ndr_referent(this.sid, 1);
+
+            if ( this.name.buffer != null ) {
+                _dst = _dst.deferred;
+                int _name_bufferl = this.name.length / 2;
+                int _name_buffers = this.name.maximum_length / 2;
+                _dst.enc_ndr_long(_name_buffers);
+                _dst.enc_ndr_long(0);
+                _dst.enc_ndr_long(_name_bufferl);
+                int _name_bufferi = _dst.index;
+                _dst.advance(2 * _name_bufferl);
+
+                _dst = _dst.derive(_name_bufferi);
+                for ( int _i = 0; _i < _name_bufferl; _i++ ) {
+                    _dst.enc_ndr_short(this.name.buffer[ _i ]);
+                }
+            }
+            if ( this.sid != null ) {
+                _dst = _dst.deferred;
+                this.sid.encode(_dst);
+
+            }
+        }
+
+
+        @Override
+        public void decode ( NdrBuffer _src ) throws NdrException {
+            _src.align(4);
+            _src.align(4);
+            if ( this.name == null ) {
+                this.name = new rpc.unicode_string();
+            }
+            this.name.length = (short) _src.dec_ndr_short();
+            this.name.maximum_length = (short) _src.dec_ndr_short();
+            int _name_bufferp = _src.dec_ndr_long();
+            int _sidp = _src.dec_ndr_long();
+
+            if ( _name_bufferp != 0 ) {
+                _src = _src.deferred;
+                int _name_buffers = _src.dec_ndr_long();
+                _src.dec_ndr_long();
+                int _name_bufferl = _src.dec_ndr_long();
+                int _name_bufferi = _src.index;
+                _src.advance(2 * _name_bufferl);
+
+                if ( this.name.buffer == null ) {
+                    if ( _name_buffers < 0 || _name_buffers > 0xFFFF )
+                        throw new NdrException(NdrException.INVALID_CONFORMANCE);
+                    this.name.buffer = new short[_name_buffers];
+                }
+                _src = _src.derive(_name_bufferi);
+                for ( int _i = 0; _i < _name_bufferl; _i++ ) {
+                    this.name.buffer[ _i ] = (short) _src.dec_ndr_short();
+                }
+            }
+            if ( _sidp != 0 ) {
+                if ( this.sid == null ) { /* YOYOYO */
+                    this.sid = new rpc.sid_t();
+                }
+                _src = _src.deferred;
+                this.sid.decode(_src);
+
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 160:195, 626:661.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.